python/lsst/daf/butler/_query.py

```diff
@@ -0,0 +1,267 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("Query",)
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, Mapping
+
+if TYPE_CHECKING:
+    from ._query_results import DataCoordinateQueryResults, DatasetQueryResults, DimensionRecordQueryResults
+    from .dimensions import DataId, DimensionElement, DimensionGroup
+    from .registry._registry import CollectionArgType
+
+
+class Query(ABC):
+    """Interface for construction and execution of complex queries."""
+
+    @abstractmethod
+    def data_ids(
+        self,
+        dimensions: DimensionGroup | Iterable[str] | str,
+        *,
+        data_id: DataId | None = None,
+        where: str = "",
+        bind: Mapping[str, Any] | None = None,
+        **kwargs: Any,
+    ) -> DataCoordinateQueryResults:
+        """Query for data IDs matching user-provided criteria.
+
+        Parameters
+        ----------
+        dimensions : `DimensionGroup`, `Dimension`, or `str`, or \
+                `~collections.abc.Iterable` [ `Dimension` or `str` ]
+            The dimensions of the data IDs to yield, as either `Dimension`
+            instances or `str`.  Will be automatically expanded to a complete
+            `DimensionGroup`.  Support for `Dimension` instances is deprecated
+            and will not be supported after v27.
```

Comment:
Looks like you can drop mention of `Dimension` instances here, at least according to the type annotation (and I endorse not putting the backwards compatibility into these new APIs).